### PR TITLE
[#11383] Workflow Automation

### PR DESCRIPTION
**Pull request by pPris**
Part of #11383 

**Brief description of PR**
* Labels newly opened PRs properly depending on if they're drafts or not
* Relabells PRs and comments on the PR with links to failing checks 
* In depth description of the functionality is a bit too long so I'm not sure if i should add it here
* Note that this pr adds two more dependencies to the repo, @actions/core and @actions/github

**Structure of the files**
* Each folder under pr_management contains a code for a different action (main.ts/main.js) and an action.yml file that contains metadata.
   * each of these folders also contain an index.js file which is bundled using @vercel/ncc. Normally for github actions, node_modules have to be checked into the repo, however, this can be avoided by using the ncc module (this was also recommended in an official github tutorial)
   * the way i built the files requires a few different commands to be run so I added some scripts to package.json
* Each yml file which is added to .github/workflows contains a workflow that runs on certain event triggers as specified in the file. (these are kept at the same level as e2e.yml, etc. because github did not recognise workflow files that aren't directly inside ./.github/workflow)
* There is a common.ts file for API requests and other helper functions that keep getting repeated.

**Demo**
https://github.com/pPris/pr-workflow-test/pull/37
https://github.com/pPris/pr-workflow-test/pull/40

**Status**
* Open to review. Maybe tsconfig.json, changes to package.json could be reviewed. Also the documentation could be looked at (if there's anything I should add so it's less confusing for future devs working on this aspect)

### CODE REVIEW

**Comment by pPris** [comment]
@bot ready for review (just testing this. at the time of comment, component & e2e tests aren't done)

**Comment by pPris** [comment]
@bot ready for review

File: .github/pr_management/common.ts
Comment (daongochieu2810) [inline .github/pr_management/common.ts:56]: We should not have  magic strings being repetitive, lets make them into const vars

File: .github/pr_management/common.ts
Comment (daongochieu2810) [inline .github/pr_management/common.ts:9]: I think its better if we abstract `github` into separate file e.g GithubManager or GithubHelper, so it will be easier to extend later

File: .github/pr_management/common.ts
Comment (daongochieu2810) [inline .github/pr_management/common.ts:12]: so for these variables, having getters for them is neater

File: .github/pr_management/common.ts
Comment (daongochieu2810) [inline .github/pr_management/common.ts:16]: I think u can use @todo tag to identify TODO task

File: .github/pr_management/common.ts
Comment (daongochieu2810) [inline .github/pr_management/common.ts:20]: These magic strings should be turned into const at the beginning of the file, so changing them in the future is easier

File: .github/pr_management/common.ts
Comment (daongochieu2810) [inline .github/pr_management/common.ts:78]: octokit can be abstracted into a class also

**Review by daongochieu2810 — CHANGES_REQUESTED** [review]

File: .github/pr_management/common.ts
Comment (pPris) [reply .github/pr_management/common.ts:20]: These strings are only used in one place in the js files, and they correspond to the action names in the yml files. Should I still abstract them into const?

**Comment by pPris** [comment]
I've seen the review, just need some time to fix this

File: .github/pr_management/common.ts
Comment (daongochieu2810) [reply .github/pr_management/common.ts:20]: @pPris I feel like its better to do so, the benefit of doing so might not be too visible now, but will be in the future

**Comment by rrtheonlyone** [comment]
@pPris @daongochieu2810 let's not go too deep into refactoring for now - it is more important to ensure the functionality is correct.

**Comment by daongochieu2810** [comment]
@pPris then I think you can ignore my comments for now, or make another issue for them

**Comment by pPris** [comment]
Sure i'll pause the refactorings and wait for comments on the functionality

File: .github/pr_management/common.ts
Comment (daongochieu2810) [inline .github/pr_management/common.ts:24]: `pr` doesnt have a type?

File: .github/pr_management/common.ts
Comment (daongochieu2810) [inline .github/pr_management/common.ts:32]: can you elaborate on this comment?

File: .github/pr_management/common.ts
Comment (daongochieu2810) [inline .github/pr_management/common.ts:45]: I think this should return a boolean, or u can change the function name

Comment (daongochieu2810) [inline .github/pr_management/common.ts:45]: I think this should return a boolean, or u can change the function name

File: .github/pr_management/common.ts
Comment (daongochieu2810) [reply .github/pr_management/common.ts:56]: @pPris actually once u have done the above, we may not need these abstractions anymore, just `addLabel` is enough

File: .github/pr_management/common.ts
Comment (daongochieu2810) [inline .github/pr_management/common.ts:138]: I think this comment might not be informative. I also notice that u have `isValidTimestamp` and `wereReviewCommentsAdded` above, which also do checkings, should we place them here instead?

**Review by daongochieu2810 — CHANGES_REQUESTED** [review]

**Comment by daongochieu2810** [comment]
@pPris I will need some time to pull your repo and test on mine

**Comment by daongochieu2810** [comment]
@pPris @rrtheonlyone the functionality seems good on my repo

**Comment by rrtheonlyone** [comment]
Thanks for checking @daongochieu2810. I

Feel free to approve on your end if you are more or less satisfied with the code. (Again, let's not spend too much time on refactoring etc. - this one we can always do later.)

File: .github/pr_management/common.ts
Comment (pPris) [reply .github/pr_management/common.ts:45]: Thank you, changing it to validateTimestamp

**Comment by hhdqirui** [comment]
@wkurniawan07 @ChooJeremy @madanalogy for your kind review. Thanks

**Review by ChooJeremy — CHANGES_REQUESTED** [review]
Apologies for the late review!

Checking your flow, you have the following logic: (correct me if I'm wrong)
1. For draft PRs, we add the ongoing label and exit
2. When a user comments ready for review, if all tests are passing, mark as to review and drop ongoing.
3. For a mon-draft PR, we check if all tests are passing
  - If all tests are passing and the PR is NOT marked as ongoing, we add the toReview label. If it is ongoing, we assume we need to wait for the user to mark as ready for review. This also means that if a PR is created with all passing tests, it is automatically marked ready for review.
  - If not all tests are passing, we notify the author if he hasn't been notified yet and mark as ongoing. Author is deemed notified if there is a notification between now and the last time the ongoing label was added. If there is no ongoing label, assume the author has not yet been notified.

I think in general this logic works. Just one case to consider:    
Typically, after a review process is complete, reviewers will mark the PR as `finalReview`. It seems like this label is currently not recognized and would cause the bot to mark it as `toReview` again. A simple way to resolve this is to add the `finalReview` label as an exception, marking it as "do not process any further". While we're at it, the same exception could be made for the `toMerge` label. If the final review fails, the PR would fall back to the ongoing label and the cycle can continue.

Ideally, we would like to add the `finalReview` label back after the author has finished making changes. A relatively simple way to do this could be to check the last added label (be it `finalReview` or `toReview` and add it back if the author marks it as ready for review) 

Other than that, I think the logic is good. There's some improvements to the code that would be good as well, particularly improvements to the naming of functions and variables, though let's focus on the functionality first.

**Comment by pPris** [comment]
Hi @ChooJeremy Thank you for pointing that case out. I've added made the changes according to your review - here's a [demo of the functionality added](https://github.com/pPris/pr-workflow-test/pull/63).

Some Technicalities: I've changed the phrase to comment to `PR ready for review` because there exists a github account called bot, and I also didn't want to remove the first word to avoid accidentally triggering the check.

The logic you outlined looks pretty much correct to me 👍

File: .github/pr_management/ready_for_review_comment/main.ts
Comment (ChooJeremy) [inline .github/pr_management/ready_for_review_comment/main.ts:32]: should we also check that the comment is by the author of the PR?

**Review by ChooJeremy — APPROVED** [review]
Functionality looks good (some minor points below).

File: .github/pr_management/ready_for_review_comment/main.ts
Comment (pPris) [reply .github/pr_management/ready_for_review_comment/main.ts:31]: This makes sense if most PRs are only worked by the author, so I've added it now

**Comment by jianhandev** [comment]
@pPris let us know once this is ready for review again, thanks.

**Comment by pPris** [comment]
@jianhandev Hi, this is ready for review, thanks

**Comment by ChooJeremy** [comment]
Logic is good. @rrtheonlyone @madanalogy would you like to review? Or perhaps we should start refactoring.

**Comment by rrtheonlyone** [comment]
Looks great @pPris ! Thanks for the detailed review @ChooJeremy 

Let me take a closer look over the coming week and see how we can get this running for us

**Comment by daongochieu2810** [comment]
@rrtheonlyone is there anything you want to add? If not I think we can merge this and get this running

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
Although it looks like it functions correctly, I'm concerned about its maintainability moving forward. For instance, the built files have to be checked in instead of built-on-demand. Without documentation or know-how, future devs will not know that you need to check in the built files instead of just the `.ts` files in order for the changes to take place.

Others that are less major (but still concerns nonetheless):
- Mixing up the files in the root `package.json` file (`@actions/*` has no business being the dependencies here)
- Missing unit tests/lints for the actions. Probably not so much on the lint part, but some form of unit tests should be there.

**Comment by rrtheonlyone** [comment]
Thanks @wkurniawan07 - and apologies on the long hiatus on review here!

I think @pPris has done a great job here and in due time, we can definitely take this forward into our repository. This will alleviate some growing pains regarding PR management.  

I have played around with the current implementation and it looks good :) I can try do a slow initial clean-up on this but have drafted by thoughts below for further discussion. We are not in a huge rush for this so we might as well get this right before merging in!

Regarding maintainability:
- we can create some kind of build workflow with existing `package.json` I think? 
- we should document this operational automation in our docs
- we can add `npm tests` for the action behaviour

Regarding functionality, it works as pointed out by early reviewers. Some observations:
- Workflow `Ready to Review` takes unusually long in some cases to execute
  -  we should be able to early comment once a single test fails and optimize the busy waiting
  -  do we really need to give a bullet list of all tests? we can maybe only show failing tests and ask them to scroll down and click details for failing test

Implementation Suggestions/Nits:
- Although it works, I don't think `await` needs to be used together with `then` (see that in some places)
- We can clean up `common.ts` for sure - perhaps using classes?
- We can name these workflows too!  (`Name: Review Label Assigner` etc.)

**Comment by pPris** [comment]
Thank you for the reviews, I agree with all the points raised. Here are some follow ups just to clarify:

1. Is there anything wrong with having to build the files and committing as long as I document this use case?

Why we have to minify the files is because GitHub Actions requires the /node_modules in .github/workflows to be commited. Minifying the files is a workaround so that we don't have to commit node_modules into this huge repo. 

Do let me know if we should just commit /node_modules instead to avoid having this manual build and making sure that devs commit it.

2. "We can clean up common.ts for sure - perhaps using classes?"

Yes I think previously dao ngoc hieu suggested abstracting away the github calls into a github manager. I'll look into how we can reorganise the api calls and other parts of the code differently. 

3. "We can create some kind of build workflow with existing package.json I think?"

There is one script that I have already added to build all three files in the respective directories, does that match what you have in mind here?

4. "Missing unit tests"

I've not figured how to go about this so let me follow up on this after the refactoring occurs.

----

Just to keep you in the loop: 

Prof's kindly extended the deadline to grade my IWM-OSS based on if this PR is merged to the end of winter vacation (early Jan). As it is finals season right now, I will be pausing work on the PR until the first week of December. Right after, I'll work on the refactor and am aiming to request a review by December week 2.

Hope that sounds alright!

**Comment by rrtheonlyone** [comment]
Thank you for the initial effort here @pPris - this is a good prototype that demonstrates how powerful Github actions can be to automate our tedious labelling efforts.

However, this is unfortunately not something we are ready to merge in at the current state. Some notable issues:
- Even though typescript is used, many of the return types are missing and not fully respected
- Code is unnecessarily verbose; much of the logic can be simplified
- There is an overabundance of unnecessary comments/abstractions everywhere (ties with the above point)
- There is no linter/tests setup (even though this is not critical: this makes it that much more harder to scale)
- I don't see why we have three separate workflows here, it can be merged into one
- You have done a good job with the documentation; however, it is not reasonable to expect everyone to know that they must first read through that file inside `.github` before doing changes

In my opinion, these files create an unnecessary pollution of our main codebase. This infrastructure-as-code is parked here in a rather odd manner.

This is not to say your efforts here are in vain! This is my proposal to use all of this.

### Proposal

We are going to create a separate repository named `teammates-actions` (I have already one here for prototyping [here](https://github.com/rrtheonlyone/teammates-actions-test). This repository
- will contain our infrastructure-as-code for our actions
- it will have main README file with clear linting/tests 

We will then publish our custom action out to the marketplace - we can even make this generic enough to be used across all our NUS OSS repositories (@damithc)

For teammates:
- we will then create exactly one workflow that uses our custom action
- we will update the testing documentation to explain how the labelling works + link to the other repo so interested parties in contribute

This way we do not unnecessarily add more complexity to this repository. It is already large enough as it is with the frontend+backend logic etc. In the long run, we can have more people contributing to the actions themselves and make it much more powerful to be used in many places.

### Plan

I have already replicated the current logic out in my prototype repo. I will still need to validate that the functionality works and add some simple unit tests to start.

Following which we can then proceed to:
- create a new repo
- update the teammates workflow files

I will update on the progress on this.


Let me know your thoughts on this! (or if you disagree with the proposal)

**Comment by samuelfangjw** [comment]
Change in project direction, will not be merged in current state.

**Pull request closed without merging**
